misk-feature/src/main/kotlin/misk/feature/MoshiExt.kt

```diff
@@ -0,0 +1,37 @@
+package misk.feature
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonDataException
+import misk.logging.getLogger
+
+private val logger = getLogger<FeatureFlags>()
+
+/**
+ * Attempts to use [JsonAdapter.failOnUnknown] and logs any issues before falling back to ignoring
+ * the unknown fields.
+ */
+fun <T> JsonAdapter<T>.toSafeJson(value: T): String {
+  return try {
+    failOnUnknown().toJson(value)
```

Comment:
Converting a `T` to JSON, with the type `T` itself defining the schema, how can there be extraneous field that necessitate `failOnUnknown()`?

Reply:
> LGTM but test cases over the failure modes would be helpful (e.g, in answering the questions I had around when a non-optional field on a `T` is not specified in the input JSON string)

@zcross tests to show what happens https://github.com/cashapp/misk/pull/1353

Reply:
you manually create the JSON in launch darkly, so you can easily add a field that doesn't exist, especially if you screw something up like `shardKey` vs `shard_key`